File: python/08_longest_palindromic_substring.py
"""
Q8. Longest Palindromic Substring (Category: Strings)

Problem: Find the longest palindromic substring in a string.

Approach: Expand-around-center. For each position try both odd and even-length
          palindromes and track the longest found.

Time Complexity : O(n²)
Space Complexity: O(1)
"""


def longest_palindrome(s: str) -> str:
    start, max_len = 0, 1

    def expand(left: int, right: int) -> None:
        nonlocal start, max_len
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        length = right - left - 1
        if length > max_len:
            max_len = length
            start = left + 1

    for i in range(len(s)):
        expand(i, i)      # Odd-length palindrome
        expand(i, i + 1)  # Even-length palindrome

    return s[start : start + max_len]


if __name__ == "__main__":
    print(longest_palindrome("babad"))  # "bab" or "aba"
    print(longest_palindrome("cbbd"))   # "bb"
    print(longest_palindrome("a"))      # "a"

File: cpp/08_LongestPalindromicSubstring.cpp
/*
 * Q8. Longest Palindromic Substring (Category: Strings)
 *
 * Problem: Find the longest palindromic substring in a string.
 *
 * Approach: Expand-around-center. For each position expand outward for both
 *           odd-length and even-length palindromes.
 *
 * Time Complexity : O(n²)
 * Space Complexity: O(1)
 */
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

int expandAroundCenter(const string& s, int left, int right) {
    while (left >= 0 && right < (int)s.length() && s[left] == s[right]) {
        left--;
        right++;
    }
    return right - left - 1; // length of palindrome
}

string longestPalindrome(string s) {
    if (s.length() < 2) return s;
    int start = 0, maxLen = 1;

    for (int i = 0; i < (int)s.length() - 1; i++) {
        int len1 = expandAroundCenter(s, i, i);     // Odd
        int len2 = expandAroundCenter(s, i, i + 1); // Even
        int len  = max(len1, len2);
        if (len > maxLen) {
            maxLen = len;
            start  = i - (len - 1) / 2;
        }
    }
    return s.substr(start, maxLen);
}

int main() {
    cout << longestPalindrome("babad") << endl; // "bab" or "aba"
    cout << longestPalindrome("cbbd")  << endl; // "bb"
    return 0;
}
